pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -257,96 +258,61 @@ class StateVectorKokkos final
-     * @param matrix Kokkos gate matrix in the device space
-     * @param wires Wires to apply gate to.
-     * @param inverse Indicates whether to use adjoint of gate.
-     */
-    void applyMultiQubitOp(const KokkosVector &matrix,
-                           const std::vector<size_t> &wires,
-                           bool inverse = false) {
-        auto &&num_qubits = this->getNumQubits();
-        if (wires.size() == 1) {
-            applySingleQubitOp(matrix, wires, inverse);
-        } else if (wires.size() == 2) {
-            applyTwoQubitOp(matrix, wires, inverse);
-        } else {
-            Kokkos::View<const size_t *, Kokkos::HostSpace,
-                         Kokkos::MemoryTraits<Kokkos::Unmanaged>>
-                wires_host(wires.data(), wires.size());
-
-            Kokkos::View<size_t *> wires_view("wires_view", wires.size());
-            Kokkos::deep_copy(wires_view, wires_host);
-
-            std::size_t two2N = std::exp2(num_qubits_ - wires.size());
-            std::size_t dim = std::exp2(wires.size());
+                two2N, apply2QubitOpFunctor<fp_t>(*data_, num_qubits,
+                                                  matrix_trans, wires));
+            break;
+        case 3:
+            Kokkos::parallel_for(
+                two2N, apply3QubitOpFunctor<fp_t>(*data_, num_qubits,
+                                                  matrix_trans, wires));
+            break;
+        case 4:
+            Kokkos::parallel_for(
+                two2N, apply4QubitOpFunctor<fp_t>(*data_, num_qubits,
+                                                  matrix_trans, wires));
+            break;
+        default:
             std::size_t scratch_size = ScratchViewComplex::shmem_size(dim) +
                                        ScratchViewSizeT::shmem_size(dim);
-
-            if (!inverse) {
-                Kokkos::parallel_for(
-                    "multiQubitOpFunctor",
-                    TeamPolicy(two2N, Kokkos::AUTO, dim)
-                        .set_scratch_size(0, Kokkos::PerTeam(scratch_size)),
-                    multiQubitOpFunctor<PrecisionT, false>(*data_, num_qubits,
-                                                           matrix, wires_view));
-            } else {
-                Kokkos::parallel_for(
-                    "multiQubitOpFunctor",
-                    TeamPolicy(two2N, Kokkos::AUTO, dim)
-                        .set_scratch_size(0, Kokkos::PerTeam(scratch_size)),
-                    multiQubitOpFunctor<PrecisionT, true>(*data_, num_qubits,
-                                                          matrix, wires_view));
-            }
+            Kokkos::parallel_for(
+                "multiQubitOpFunctor",
+                TeamPolicy(two2N, Kokkos::AUTO, dim)
+                    .set_scratch_size(0, Kokkos::PerTeam(scratch_size)),
+                multiQubitOpFunctor<PrecisionT>(*data_, num_qubits,
+                                                matrix_trans, wires));
```

Comment:
We don't need to worry about the inverse at this point anymore?

Reply:
I'm assuming unitary gates, and hence the inverse is obtained with the `parallel_for` above taking the transpose conjugate. We have to do this once, on a relatively small array. This way, we don't need to template each functor on `inverse` and (almost) duplicate the body (see [here](https://github.com/PennyLaneAI/pennylane-lightning/pull/490/files#diff-5e7d16242d2c553f5c54d49701599fb90cf2157979006d8e3bd29c845035f9cbL139) for example)). I'm not sure whether the compiler used to optimize that away, but now we don't need to take the conjugate transpose of the matrix on the fly, and hence we always get contiguous access during the gate application.

Reply:
Got it, thanks!